dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs

```diff
@@ -0,0 +1,158 @@
+
+        // Get the chat completion results
+        var completionResults = await chatCompletion.GetChatCompletionsAsync(chatMessages, cancellationToken: cancellationToken).ConfigureAwait(false);
+        var completionMessage = await completionResults[0].GetChatMessageAsync(cancellationToken).ConfigureAwait(false);
+
+        var resultContext = this._kernel.CreateNewContext();
+        resultContext.Variables.Update(completionMessage.Content);
+
+        Match match = Regex.Match(resultContext.Result, @"````\\s*(handlebars)?\\s*(.*?)\\s*```", RegexOptions.Singleline);
+        if (!match.Success)
+        {
+            throw new SKException("Could not find the plan in the results");
+        }
+
+        var template = match.Groups[2].Value.Trim(); // match.Success ? match.Groups[2].Value.Trim() : resultContext.Result;
+
+        template = template.Replace("compare.equal", "equal");
+        template = template.Replace("compare.lessThan", "lessThan");
+        template = template.Replace("compare.greaterThan", "greaterThan");
+        template = template.Replace("compare.lessThanOrEqual", "lessThanOrEqual");
+        template = template.Replace("compare.greaterThanOrEqual", "greaterThanOrEqual");
+        template = template.Replace("compare.greaterThanOrEqual", "greaterThanOrEqual");
+
+        template = MinifyHandlebarsTemplate(template);
+        return new HandlebarsPlan(this._kernel, template);
+    }
+
+    private List<FunctionView> GetAvailableFunctionsManual(CancellationToken cancellationToken = default)
+    {
+        return this._kernel.Functions.GetFunctionViews()
+            .Where(s => !this._config.ExcludedPlugins.Contains(s.PluginName, StringComparer.OrdinalIgnoreCase)
+                && !this._config.ExcludedFunctions.Contains(s.Name, StringComparer.OrdinalIgnoreCase)
+                && !s.Name.Contains("Planner_Excluded"))
+            .ToList();
+    }
+
+    private ChatHistory GetChatHistoryFromPrompt(string prompt, IChatCompletion chatCompletion)
+    {
+        // Extract the chat history from the rendered prompt
+        string pattern = @"<(user~|system~|assistant~)>(.*?)<\/\1>";
+        MatchCollection matches = Regex.Matches(prompt, pattern, RegexOptions.Singleline);
+
+        // Add the chat history to the chat
+        ChatHistory chatMessages = chatCompletion.CreateNewChat();
+        foreach (Match m in matches.Cast<Match>())
+        {
+            string role = m.Groups[1].Value;
+            string message = m.Groups[2].Value;
+
+            switch (role)
+            {
+                case "user~":
+                    chatMessages.AddUserMessage(message);
+                    break;
+                case "system~":
+                    chatMessages.AddSystemMessage(message);
+                    break;
+                case "assistant~":
+                    chatMessages.AddAssistantMessage(message);
+                    break;
```

Comment:
default case?

Reply:
A default case is not needed here because we don't want to parse out any other kind of messages except these certain ones. Others would be considered a failed case and cause the template to throw an exception when rendered.